### PR TITLE
return the current rebalance result if already done

### PR DESCRIPTION
**Pull request by jadami10**
This is a `bugfix` on top of #13281 to better harden returning the rebalance result.

#13281 will likely fix the same issue, but doesn't prevent the fact that we can just forget to persist/update the job id in ZK. Since we're already waiting for the job id to persist, this PR attempts to return the actual rebalance result if it's already complete. Since Pinot 1.1 still has bugs with job id not being tracked correctly, we'll be using this internally to limit the number of times we see polling job ids fail.

### CODE REVIEW

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTableRestletResource.java [hunk 1]
```diff
@@ -677,14 +678,30 @@ public RebalanceResult rebalance(
...
+              return new RebalanceResult(rebalanceJobId, RebalanceResult.Status.FAILED, errorMsg, null, null, null);
             }
           });
-          waitForJobIdToPersist(dryRunResult.getJobId(), tableNameWithType);
+          boolean isJobIdPersisted = waitForJobIdToPersist(dryRunResult.getJobId(), tableNameWithType);
+
+          // It's possible the dryRun indicates a rebalance is needed, but next rebalance is not.
+          // In that case, we never persist the jobId, but will already have a rebalanceResult.
+          // Return the actual rebalanceResult in this case.
+          if (!isJobIdPersisted && rebalanceResultFuture.isDone()) {
```
Comment (jadami10) [inline pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTableRestletResource.java:696]: I did this mainly to try to preserve behavior. But we could also only check `rebalanceResultFuture.isDone()`

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTableRestletResource.java [hunk 2]
```diff
@@ -677,14 +678,30 @@ public RebalanceResult rebalance(
         if (dryRunResult.getStatus() == RebalanceResult.Status.DONE) {
           // If dry-run succeeded, run rebalance asynchronously
           rebalanceConfig.setDryRun(false);
-          _executorService.submit(() -> {
+          Future<RebalanceResult> rebalanceResultFuture = _executorService.submit(() -> {
             try {
-              _pinotHelixResourceManager.rebalanceTable(tableNameWithType, rebalanceConfig, rebalanceJobId, true);
+              return _pinotHelixResourceManager.rebalanceTable(
+                  tableNameWithType, rebalanceConfig, rebalanceJobId, true);
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTableRestletResource.java:684]: (minor) This reformat seems unnecessary

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTableRestletResource.java [hunk 3]
```diff
@@ -677,14 +678,30 @@ public RebalanceResult rebalance(
...
+
+          // It's possible the dryRun indicates a rebalance is needed, but next rebalance is not.
+          // In that case, we never persist the jobId, but will already have a rebalanceResult.
+          // Return the actual rebalanceResult in this case.
+          if (!isJobIdPersisted && rebalanceResultFuture.isDone()) {
+            try {
+              return rebalanceResultFuture.get();
+            } catch (Throwable ignored) {
+              // Ignore any errors here. Errors will have already been logged
+              // in the submitted future.
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTableRestletResource.java:701]: (minor) Wrong formatting

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTableRestletResource.java (hunk 1 above)
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTableRestletResource.java:696]: Should we consider always return `rebalanceResultFuture.get()` if it is already done?

**Review by Jackie-Jiang — COMMENTED** [review]
In `waitForJobIdToPersist()`, we can consider passing in the future and stop waiting if the future is done. This way we can avoid unnecessary wait because job id will never show up

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTableRestletResource.java (hunk 2 above)
Comment (jadami10) [reply pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTableRestletResource.java:684]: I don't know why, but `mvn checkstyle:check` complains without this

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTableRestletResource.java (hunk 1 above)
Comment (jadami10) [reply pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTableRestletResource.java]: ya, i figured I'd leave the old behavior. But we might as well just return if done.

**Comment by jadami10** [comment]
> In `waitForJobIdToPersist()`, we can consider passing in the future and stop waiting if the future is done. This way we can avoid unnecessary wait because job id will never show up

good point. added that check as well

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTableRestletResource.java (hunk 2 above)
Comment (Jackie-Jiang) [reply pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTableRestletResource.java:684]: You are right, I didn't realize there is an extra `return` added to the line :-P

**Review by Jackie-Jiang — APPROVED** [review]

**Pull request merged**